File: jfix-zookeeper/src/main/java/ru/fix/zookeeper/lock/LockData.java
package ru.fix.zookeeper.lock;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;
import org.slf4j.Logger;

import java.net.InetAddress;
import java.net.UnknownHostException;

@Data
class LockData {

    @JsonProperty
    private String ip = "Unknown ip";

    @JsonProperty
    private String hostname = "Unknown hostname";

    @JsonProperty
    private String serverId;

    @JsonProperty
    private String uuid;

    @JsonProperty
    private long expirationDate;

    @JsonCreator
    LockData(@JsonProperty("ip") String ip, @JsonProperty("hostname") String hostname,
                    @JsonProperty("serverId") String serverId, @JsonProperty("uuid") String uuid,
                    @JsonProperty("expirationDate") long expirationDate) {
        this.ip = ip;
        this.hostname = hostname;
        this.serverId = serverId;
        this.uuid = uuid;
        this.expirationDate = expirationDate;
    }

    LockData(String uuid, long expirationDate, String serverId, Logger logger) {
        this.uuid = uuid;
        this.expirationDate = expirationDate;
        this.serverId = serverId;

        try {
            InetAddress inetAddr = InetAddress.getLocalHost();
            this.ip = inetAddr.getHostAddress();
            this.hostname = inetAddr.getHostName();
        } catch (UnknownHostException ignore) {
            logger.trace("Node[serverId={}, uuid={}] already removed on release.", serverId, uuid, ignore);
        }
    }
}

File: jfix-zookeeper/src/main/kotlin/ru/fix/zookeeper/lock/LockData.kt
package ru.fix.zookeeper.lock

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonIgnore
import org.slf4j.LoggerFactory
import java.net.InetAddress
import java.net.UnknownHostException
import java.time.Duration
import java.time.Instant

data class LockData(
        val uuid: String,
        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern ="yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "UTC")
        val expirationTimestamp: Instant,
        val metadata: String? = null
) {
    var ip: String = "Unknown ip"
        private set
    var hostname: String = "Unknown hostname"
        private set

    companion object {
        private val logger = LoggerFactory.getLogger(LockData::class.java)
    }

    init {
        try {
            val inetAddress = InetAddress.getLocalHost()
            ip = inetAddress.hostAddress
            hostname = inetAddress.hostName
        } catch (e: UnknownHostException) {
            logger.debug("Host is invalid when lock data instantiated", e)
        }
    }

    constructor(
            ip: String,
            hostname: String,
            uuid: String,
            expirationDate: Instant
    ) : this(uuid, expirationDate) {
        this.ip = ip
        this.hostname = hostname
    }

    @JsonIgnore
    fun isExpiredWithTimeout(timeout: Duration): Boolean {
        return expirationTimestamp.plus(timeout).isBefore(Instant.now())
    }

    @JsonIgnore
    fun isExpired(): Boolean {
        return expirationTimestamp.isBefore(Instant.now())
    }

    @JsonIgnore
    fun isOwnedBy(uuid: String): Boolean {
        return this.uuid == uuid
    }
}
